File: composite-pattern/organization/organization.ts
// Component
interface EmployeeComponent {
  showDetails(): void;
}

// Leaf
class Employee implements EmployeeComponent {
  private name: string;
  private position: string;

  constructor(name: string, position: string) {
    this.name = name;
    this.position = position;
  }

  showDetails(): void {
    console.log(`${this.position} :: ${this.name}`);
  }
}

// Composite
class Manager implements EmployeeComponent {
  private name: string;
  private position: string;
  private subordinates: EmployeeComponent[] = [];

  constructor(name: string, position: string) {
    this.name = name;
    this.position = position;
    this.subordinates = [];
  }

  add(employee: EmployeeComponent): void {
    this.subordinates.push(employee);
  }

  remove(employee: EmployeeComponent): void {
    this.subordinates = this.subordinates.filter((e) => e !== employee);
  }

  showDetails(): void {
    console.log(`${this.position} -- ${this.name}`);

    for (const subordinate of this.subordinates) {
      subordinate.showDetails();
    }
  }
}

// Usage
const emp1 = new Employee("Shreyash", "Senior Engineer");
const emp2 = new Employee("Jenny", "SEO expert");

const manager = new Manager("Mahesh", "Tech lead");
manager.add(emp1);
manager.add(emp2);

manager.showDetails();

File: composite-pattern/organization/organization.rs
use std::rc::Rc;

// Component
trait EmployeeComponent {
    fn show_details(&self);
}

// Leaf
struct Employee {
    name: String,
    position: String,
}

impl EmployeeComponent for Employee {
    fn show_details(&self) {
        println!("Position: {}, Name: {}", self.position, self.name);
    }
}

impl Employee {
    fn new(name: &str, position: &str) -> Self {
        Self {
            name: name.to_string(),
            position: position.to_string(),
        }
    }
}

// Composite
struct Manager {
    name: String,
    position: String,
    subordinates: Vec<Rc<dyn EmployeeComponent>>,
}

impl EmployeeComponent for Manager {
    fn show_details(&self) {
        println!("Position: {}, Name: {}", self.position, self.name);

        for subord in &self.subordinates {
            subord.show_details();
        }
    }
}

impl Manager {
    fn new(name: &str, position: &str) -> Self {
        Self {
            name: name.to_string(),
            position: position.to_string(),
            subordinates: vec![],
        }
    }

    fn add(&mut self, employee: Rc<dyn EmployeeComponent>) {
        self.subordinates.push(employee);
    }
}

fn main() {
    let emp1 = Rc::new(Employee::new("Shreyash", "ML Engineer"));
    let emp2 = Rc::new(Employee::new("Khushboo", "Software Engineer"));

    let mut manager = Manager::new("Gaurav", "Staff Engineer");
    manager.add(emp1.clone());
    manager.add(emp2.clone());

    manager.show_details();
}
